crates/noirc_evaluator/src/ssa/flatten.rs

```diff
@@ -373,53 +377,110 @@ fn evaluate_object(
     }
 }
 
-pub fn inline_tree(ctx: &mut SsaContext) {
+//inline main
+pub fn inline_tree(ctx: &mut SsaContext, block_id: BlockId) {
     //inline all function calls
-    inline_block(ctx, ctx.first_block);
+    let mut retry = MAX_INLINE_TRIES;
+    while retry > 0 && !inline_block(ctx, ctx.first_block) {
+        retry -= 1;
+    }
+    assert!(retry > 0, "Error - too many nested calls");
+    for b in ctx[block_id].dominated.clone() {
+        inline_tree(ctx, b);
+    }
+}
+
+fn inline_cfg(ctx: &mut SsaContext, entry_block: BlockId) -> bool {
+    let mut result = true;
+    let func_cfg = block::bfs(entry_block, None, ctx);
+    for block_id in func_cfg {
+        if !inline_block(ctx, block_id) {
+            result = false;
+        }
+    }
+    result
+}
+
+//inline each function so that a function does not call anyy other function
+//There is probably an optimal order to use when inlining the functions
+//for now we inline all functions one time, and then recurse until no more function call
+pub fn inline_all_functions(ctx: &mut SsaContext) {
+    let mut nested_call = true;
+    let mut retry = MAX_INLINE_TRIES;
+    let func_cfg: Vec<BlockId> =
+        noirc_frontend::util::vecmap(ctx.functions.values(), |f| f.entry_block);
```

Comment:
Nitpick, but lets import vecmap here to reduce the code noise a bit

Reply:
Done.